File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        i = 0
        
        def helper(hi: int) -> Optional[TreeNode]:
            nonlocal i
            if i >= len(preorder) or preorder[i] > hi:
                return None
            root = TreeNode(preorder[i])
            i += 1
            root.left = helper(root.val)
            root.right = helper(hi)
            return root
        
        return helper(float("inf"))

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @return {TreeNode}
 */
var bstFromPreorder = function(preorder) {
    let i = 0
    return helper(Infinity)
    
    function helper(hi) {
        if (i >= preorder.length || preorder[i] > hi) return null
        const root = new TreeNode(preorder[i])
        i++
        root.left = helper(root.val)
        root.right = helper(hi)
        return root
    }
};